### PR TITLE
Shared BreadCrumb

### PR DESCRIPTION
**Pull request by cameron-freshworks**
*Issue #:* /bcgov/entity#10232

*Description of changes:*
* Implements a simple shared BreadCrumb
* Added BreadCrumb interface
* BreadCrumb Stories
* Unit tests
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the business-filings-ui license (Apache 2.0).

### CODE REVIEW

**Comment by cameron-freshworks** [comment]
> _Issue #:_ /[bcgov/entity#10232](https://github.com/bcgov/entity/issues/10232)
> 
> _Description of changes:_
> 
> * Implements Shared BreadCrumb
> * Added BreadCrumb interface
> * BreadCrumb Stories
> 
> By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the business-filings-ui license (Apache 2.0).

**Comment by severinbeauvais** [comment]
Could you please attach some screen shots of the different variations? (So we don't have to build locally to test ;) )

**Comment by cameron-freshworks** [comment]
> Could you please attach some screen shots of the different variations? (So we don't have to build locally to test ;) )

Totally, can do! I wish i could share it in Storybook as a preview some how...

File: src/components/BreadCrumb/BreadCrumb.vue
Comment (severinbeauvais) [inline src/components/BreadCrumb/BreadCrumb.vue:66]: What if no match is found?

**Comment by cameron-freshworks** [comment]
Screen Shot 2021-12-17 at 11 17 09 AM
Screen Shot 2021-12-17 at 11 17 03 AM
Screen Shot 2021-12-17 at 11 16 52 AM
Screen Shot 2021-12-17 at 11 16 27 AM
Screen Shot 2021-12-17 at 11 16 20 AM
Screen Shot 2021-12-17 at 11 16 15 AM

File: src/interfaces/breadcrumb-interface.ts
Comment (severinbeauvais) [inline src/interfaces/breadcrumb-interface.ts:6]: Why are these optional? Looks like the template code doesn't handle these undefined properties.

File: src/components/BreadCrumb/BreadCrumb.vue
Comment (severinbeauvais) [inline src/components/BreadCrumb/BreadCrumb.vue:116]: Please add a unit test for this.

File: src/components/BreadCrumb/BreadCrumb.vue
Comment (cameron-freshworks) [reply src/components/BreadCrumb/BreadCrumb.vue:66]: Then something is broken :) 
Without the route the user is most likely not in the application any longer.
I suppose it is a requirement to have a `to` on local starting point, so we can count on that.
`to` is optional according to the interface as we will be passing external crumbs which rely on a href, but locally (in our ui's) we will always have/need `to` routes for local navigation. 

So tldr: this check finds our starting point, regardless of how many href crumbs before our local starting point and subsequently tracks any deeper navigation through the `to`. Although we don't have too many implementations of deeper navigation outside the filings ui.

**Comment by cameron-freshworks** [comment]
> Could you please attach some screen shots of the different variations? (So we don't have to build locally to test ;) )

See pics below!

File: src/interfaces/breadcrumb-interface.ts
Comment (cameron-freshworks) [reply src/interfaces/breadcrumb-interface.ts:6]: Well, a crumb can have either a to or a href, neither are required although 1 is. 
Not sure how to type that.

File: src/components/BreadCrumb/BreadCrumb.vue
Comment (cameron-freshworks) [reply src/components/BreadCrumb/BreadCrumb.vue:116]: Yea, will do. thinking on what to test here that is meaningful.

File: src/components/BreadCrumb/BreadCrumb.vue
Comment (severinbeauvais) [reply src/components/BreadCrumb/BreadCrumb.vue:66]: OK. So we leave it up to the user of this component to configure it correctly or else.

Is there any risk of the configuration being messed up under normal running conditions (since I know it's dynamic)?

**Comment by severinbeauvais** [comment]
> > Could you please attach some screen shots of the different variations? (So we don't have to build locally to test ;) )
> 
> See pics below!

Looks amazing!

File: src/components/BreadCrumb/BreadCrumb.vue
Comment (cameron-freshworks) [reply src/components/BreadCrumb/BreadCrumb.vue:66]: Well, it's no real risk. 
It will simply exist but not do anything. 
For example, if a user does not pass href or to values, the navigation will not go anywhere. 
If they don't pass in any text, there will be crumb on display and no link etc. 

So essentially, no real risk, any persons implementing this can see the example crumb on the prop or refer to the interface. Functionally it just won't do what they want it to do until they pass it a value.

File: src/interfaces/breadcrumb-interface.ts
Comment (severinbeauvais) [reply src/interfaces/breadcrumb-interface.ts:6]: Not easily/cleanly. I see you just pass those into the breacrumb item, so let's ignore this for now.

Are `disabled` and `exact` used atm?

File: src/components/BreadCrumb/BreadCrumb.vue
Comment (severinbeauvais) [reply src/components/BreadCrumb/BreadCrumb.vue:66]: I buy that explanation - thanks.

One thing to watch for is that an unhandled exception can cause a method/component to stop running/displaying. This could be serious depending on what gets broken (imagine something in App or TodoList or ...). In this case I think the risk is low: this component may "crash" but I don't think it will take anything down with it, so I think we're good.

File: src/interfaces/breadcrumb-interface.ts
Comment (cameron-freshworks) [reply src/interfaces/breadcrumb-interface.ts:6]: Not here but i saw it passed in some of the ui's.

File: src/components/BreadCrumb/BreadCrumb.vue
Comment (cameron-freshworks) [reply src/components/BreadCrumb/BreadCrumb.vue:116]: Ok, thought of some, implementing!

File: src/interfaces/breadcrumb-interface.ts
Comment (severinbeauvais) [reply src/interfaces/breadcrumb-interface.ts:6]: If there is no requirement to use disabled and exact then I suggest deleting them.

Yes, this would break UIs that think these do something, but on the other hand those UIs will need a bit of refactoring to use this shared component anyway. I think it would be better to not carry forward properties that do nothing.

Comment (cameron-freshworks) [reply src/interfaces/breadcrumb-interface.ts:6]: Ok right on, will do!

**Review by lekshmimallika-aot — APPROVED** [review]

File: src/components/BreadCrumb/BreadCrumb.stories.ts
Comment (severinbeauvais) [inline src/components/BreadCrumb/BreadCrumb.stories.ts:23]: Obsolete property. Also below.

File: tests/unit/BreadCrumb.spec.ts
Comment (severinbeauvais) [inline tests/unit/BreadCrumb.spec.ts:16]: Obsolete property. Also below.

**Review by severinbeauvais — APPROVED** [review]
Looks good. Nice tests.

I pre-approve so you can merge after remaining comments are resolved.

File: tests/unit/BreadCrumb.spec.ts
Comment (cameron-freshworks) [reply tests/unit/BreadCrumb.spec.ts:16]: thanks, will do!

**Pull request merged**
